Implement reference for lerp

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87182
* #87183
* #87205
* #87204
* #87203
* __->__ #87424

We follow the vectorised CPU implementation for numerical accuracy

cc @ezyang @mruberry @ngimel @Lezcano @fdrocha

lezcano committed:
Implement reference for lerp

We follow the vectorised CPU implementation for numerical accuracy

[ghstack-poisoned]

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -4025,6 +4026,29 @@ def arange(
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+)
+def lerp(start: Tensor, end: Tensor, weight: Union[Tensor, NumberType]):
+    check(start.dtype == end.dtype, lambda: f"expected dtype {start.dtype} for `end` but got dtype {end.dtype}")
```
too late for that after type promotion wrapper?

lezcano replied:
I see this as "the reference implements an extended behaviour over the PyTorch op", as we support a few more combinations of dtypes.

ngimel replied:
Ok, but this check in reference is then pointless, the inputs it's getting are all the same type.

lezcano replied:
ah, I see what you mean. I'll clean this up later in this stack.

lezcano replied:
Added a clean-up in https://github.com/pytorch/pytorch/pull/87183

lezcano committed:
Update on "Implement reference for lerp"


We follow the vectorised CPU implementation for numerical accuracy

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Implement reference for lerp"


We follow the vectorised CPU implementation for numerical accuracy

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Implement reference for lerp"


We follow the vectorised CPU implementation for numerical accuracy

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Implement reference for lerp"


We follow the vectorised CPU implementation for numerical accuracy

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Implement reference for lerp"


We follow the vectorised CPU implementation for numerical accuracy

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

ezyang commented on torch/_refs/__init__.py:
```diff
@@ -4060,6 +4061,36 @@ def arange(
+        )
+    assert isinstance(weight, Tensor)  # mypy
+    # We implement it this way for numerical stability. We assume (in the stability optimisation)
+    # that 0 <= weight <= 1. We take the abs to deal with comples numbers
```
Complex

lezcano commented:
@pytorchbot merge

I'll fix the errata in the next PR of the stack
